### ISSUE TITLE
_c_urlarg.c tests should be in the core, not in web

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4162]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| spiv reported|
|-|-|
|Trac ID|trac#4162|
|Type|defect|
|Created|2009-12-09 22:25:48Z|
|Branch|https://github.com/twisted/twisted/tree/curlarg-eliminate-4162|

`_c_urlarg.c` is part of the core (it is in `twisted/protocols`), but its unit tests are in `twisted/web/test/test_http.py`.  Those tests should be in the core somewhere instead.

This was originally mentioned in [#4142](https://github.com/twisted/twisted/issues/4142).

<details><summary>Searchable metadata</summary>

```
trac-id__4162 4162
type__defect defect
reporter__spiv spiv
priority__highest highest
milestone__ 
branch__branches_curlarg_eliminate_4162 branches/curlarg-eliminate-4162
branch_author__PenguinOfDoom PenguinOfDoom
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1260397548000000 1260397548000000
changetime__1300306459000000 1300306459000000
version__None None
owner__ 
cc__jesstess cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|spiv commented|
|-|-|

As idnar commented on IRC, another option might be to move _c_urlarg itself to web.  I'm don't have a strong opinion either way.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to PenguinOfDoom|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|PenguinOfDoom commented|
|-|-|

_c_urlarg is used by mail as well as web, so core is the place for it.

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|pahan commented|
|-|-|

(In [28754]) Branching to 'curlarg-tests-4162'

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|PenguinOfDoom removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to PenguinOfDoom|
|-|-|

You forgot to have some revisions: http://twistedmatrix.com/trac/log/branches/curlarg-tests-4162

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|PenguinOfDoom removed owner|
|-|-|

Ha ha look how dumb I am. I committed some changes to the branch now.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_10"></a>|@jesstess set owner to PenguinOfDoom|
|-|-|

Thanks for working on this, PenguinOfDoom. Some feedback:
 * I don't know if being a C file changes how this works, but presumably _c_urlarg.py needs  something like /* -*- test-case-name: twisted.protocol.test.test_c_urlarg -*- */ at the top. If that gets added, the file needs a copyright bump.
 * twisted/web/test/test_http.py has pyflakes warnings about the unused urllib import and the unused transport on line 1391.
 * test_http.py has one spot with trailing whitespace.
 * the tests in test_c_urlarg.py need have names of the form test_myTest, and should be using assert forms instead of fail forms as per the [test standard](http://twistedmatrix.com/documents/current/core/development/policy/test-standard.html). The docstrings on the tests are vague and aren't actually describing the behavior we should be seeing, and testUnquote is testing multiple behaviors, which should really be split into their own tests. Can you fix these things here or open a new ticket for them?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra set owner to @cyli|
|-|-|

let me know if the branch needs to be merged forward.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_12"></a>|@cyli set owner to PenguinOfDoom|
|-|-|

This looks like it's waiting for the author (PenguinOfDoom) to respond to review feedback - not sure why it was assigned to me.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|pahan commented|
|-|-|

(In [31154]) Branching to 'curlarg-eliminate-4162'

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|PenguinOfDoom commented|
|-|-|

There is no credible evidence that _c_urlarg ever sped anything up, and it certainly has caused its share of problems over the years. Here's a patch to delete it!

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh set owner to @lvh<br>@lvh set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh set owner to PenguinOfDoom<br>@lvh set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_18"></a>|@lvh commented|
|-|-|

Apparently there is a failing test, but I don't see how it's related...

[http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/curlarg-eliminate-4162]

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|PenguinOfDoom removed owner|
|-|-|

Oops. Fixed!

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_20"></a>|@lvh commented|
|-|-|

Looks good! Merge ftw

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|pahan set status to `closed`|
|-|-|

(In [31167]) Merge curlarg-eliminate-4162: Remove _c_urlarg

Author: PenguinOfDoom
Reviewer: lvh
Fixes: [#4162](https://github.com/twisted/twisted/issues/4162)

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|PenguinOfDoom commented|
|-|-|

[#4246](https://github.com/twisted/twisted/issues/4246) is a duplicate of this
